doc/examples/nn_regression/nn_regression.cpp

```diff
@@ -0,0 +1,409 @@
+ *
+ * \brief MLPACK TUTORIAL: neural network regression
+ * \details Real world example which shows how to create a neural network mlpack/C++ model for regression,
+ * how to save and load the model and then use it for prediction (inference).
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/data/split_data.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+#include <mlpack/core/optimizers/adam/adam_update.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+#include <mlpack/prereqs.hpp>
+
+using namespace arma;
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+
+//TODO(zso): this should go into mlpack/core/data/feature_normalization.hpp
+namespace mlpack {
+	namespace data {
+
+		/** 
+		* \class FeatureNormalization
+		* \brief Data normalization extension module for mlpack.
+		* 
+		* \details Normalization of all features in the input data into the same range provides in most of the cases
+	    * a much better accuracy and smaller and faster models! Subsequent normalizations of new data for the same model
+		* must use the same min, max vectors and therefore they are saved in the FeatureNormalization class! You can also
+		* serialize the FeatureNormalization object and reconstruct the min, max vectors e.g. for later inference/prediction.
+		*
+		* @author Zoltan Somogyi
+		*/
+		class FeatureNormalization
+		{
+		public:
+			/** Constructor */
+			FeatureNormalization(double scalelower, double scaleupper, arma::vec & _itemMin, arma::vec & _itemMax)
+				: scale_upper(scaleupper), scale_lower(scalelower), itemMin(&_itemMin), itemMax(&_itemMax) {}
+
+			/**
+			* \fn Normalize
+			*
+			* Normalize the data into the range provided by the scale_lower and scale_upper parameters 
+			* by using the itemMin and itemMax vectors or if they do not exist yet then calculate 
+			* itemMin and itemMax first for the input data.
+			* The operation is in place (modifies the input data).
+			*
+			* @param data Input dataset
+			* @see arma::mat in mlpack documentation
+			*/
+			void Normalize(arma::mat& data)
+			{
+				// Use the itemMin, itemMax vectors if provided (e.g. for subsequent normalization with the same range)
+				// or make them if not provided
+				if (itemMin->size() < 1 || itemMax->size() < 1) {
+					*itemMin = arma::vec(data.n_rows, arma::fill::zeros);
+					*itemMax = arma::vec(data.n_rows, arma::fill::zeros);
+					for (int i = 0; i < data.n_rows; i++)
+					{
+						(*itemMin)(i) = arma::min(data.row(i));
+						(*itemMax)(i) = arma::max(data.row(i));
+					}
```

Comment:
Because Armadillo is column-major, I think it will be much faster to do this in a vectorized operation also (which will iterate over columns not rows):

```
itemMin = arma::min(data, 1);
itemMax = arma::max(data, 1);
```